lib/dataset-tools/evaluation/sentence_evaluator.js

```diff
@@ -330,128 +332,141 @@ class CollectSentenceStatistics extends Stream.Writable {
         super({ objectMode: true });
 
         this._maxComplexity = options.maxComplexity;
+        this._splitByDevice = options.splitByDevice;
 
-        this._buffer = {
-            total: 0,
-            primitives: 0,
-            compounds: 0,
-            ok: [],
-            ok_without_param: [],
-            ok_function: [],
-            ok_device: [],
-            ok_num_function: [],
-            ok_syntax: [],
-            'prim/ok': [],
-            'prim/ok_without_param': [],
-            'prim/ok_function': [],
-            'prim/ok_device': [],
-            'prim/ok_num_function': [],
-            'prim/ok_syntax': [],
-            'comp/ok': [],
-            'comp/ok_without_param': [],
-            'comp/ok_function': [],
-            'comp/ok_device': [],
-            'comp/ok_num_function': [],
-            'comp/ok_syntax': [],
-        };
+        // _buffer will map devices to individual buffers. If splitByDevice is
+        // false, there will only be a single buffer. Othersie, these individual
+        // buffers will be created ad-hoc, as new devices come up.
+        this._buffer = {};
     }
 
     _write(ex, encoding, callback) {
-        this._buffer.total ++;
+        const uniqueDevices = ex.target_devices.filter((val, ix) => ex.target_devices.indexOf(val) === ix);
+        assert(uniqueDevices.length === 1);
```

Comment:
I don't understand what this code is trying to do?
Why do you not loop over all devices? It seems it would solve all the problems you were mentioning in the cover letter.

Reply:
I think you are right. I think I was too heavily anchored on the idea that this was hacky (since when I was experimenting with it, I was only getting a single string here).

The idea behind this code was that I wanted to filter out duplicates. I think this was misguided because it seems like this wasn't an array of devices, like I thought it was, but rather a singleton array with a string that had all "devices" (loosely defined - "transaction.execute()" was showing up here as a device for example. At least that's what seemed to be showing up in my tests. But you're right, a clean loop here would solve it. I'll get to it.

Reply:
It comes from goldDevices, which should be an array of device names. It will contain duplicates, but most of the examples only have one device. You can dedupe with `Array.from(new Set(...))`
Now, the problem might be in requoting.js where getFunctions and getDevices are defined. They consider all '@' tokens to be function calls, which was the case before we introduced dialogue states but is not correct now. We should fix that.

Reply:
Looking into that! Thanks!